ledger/src/blockstore.rs

```diff
@@ -3228,14 +3228,13 @@ impl Blockstore {
     // Returns the existing shred if `new_shred` is not equal to the existing shred at the
     // given slot and index as this implies the leader generated two different shreds with
     // the same slot and index
-    pub fn is_shred_duplicate(&self, shred: ShredId, payload: Vec<u8>) -> Option<Vec<u8>> {
-        let (slot, index, shred_type) = shred.unpack();
+    pub fn is_shred_duplicate(&self, new_shred: &Shred) -> Option<Vec<u8>> {
+        let (slot, index, shred_type) = new_shred.id().unpack();
         let existing_shred = match shred_type {
             ShredType::Data => self.get_data_shred(slot, index as u64),
             ShredType::Code => self.get_coding_shred(slot, index as u64),
         }
         .expect("fetch from DuplicateSlots column family failed")?;
-        let new_shred = Shred::new_from_serialized_shred(payload).unwrap();
         (existing_shred != *new_shred.payload()).then_some(existing_shred)
```

Comment:
This is the right thing to do but iirc we do modify the payload before writing to blockstore.

Shouldn't this here _also_ compare against `bytes_to_store`?
https://github.com/solana-labs/solana/blob/3b8bcb4f7/ledger/src/shred.rs#L446

Reply:
> but iirc we do modify the payload before writing to blockstore.

Correct, we do. But, we pad it out back out on retrieval (also recall that the trim/re-pad is only on data shreds):
https://github.com/solana-labs/solana/blob/a8b0c92873e4d42d4e28b5bdff14a1aca9268bdd/ledger/src/blockstore.rs#L1629-L1631

That being said, we do eventually turn the `existing_shred` byte vector into a `Shred` here:
https://github.com/solana-labs/solana/blob/a8b0c92873e4d42d4e28b5bdff14a1aca9268bdd/gossip/src/duplicate_shred.rs#L132

Kind of leaning towards just turning `existing_shred` into a `Shred` in this function; walking the call-stack, there is a bit of mismatch to see `shred` as a `Shred` but `other_payload` as a `Vec<u8>`.
https://github.com/solana-labs/solana/blob/a8b0c92873e4d42d4e28b5bdff14a1aca9268bdd/gossip/src/crds_gossip.rs#L95-L96

Agreed ?

Reply:
> Kind of leaning towards just turning existing_shred into a Shred here in Blockstore::is_shred_duplicate(); walking the call-stack, there is a bit of mismatch to see shred as a Shred but other_payload as a Vec<u8>.

I think the goal here was to not unnecessarily deserialize shreds read from blockstore.